### PR TITLE
Rework types calculation with locks

### LINKED ISSUE #152: 3.9.1 performance degradation
The thread-safety fix #151 causes performance degradation, due to lack of memoization

### PR DESCRIPTION
**Pull request by ameyer-rigetti**
Reintroduces memoization for `Message.types()` but now with thread locking.

Closes #152

### CODE REVIEW

File: rpcq/_base.py [hunk 1]
```diff
@@ -109,7 +110,8 @@ def __eq__(self, other):
     def __hash__(self):
         return hash((self.__class__, astuple(self)))
 
-    _types = {}
+    _types = None
+    _types_lock = threading.Lock()
 
     @staticmethod
     def types():
```
Comment (dbanty) [inline rpcq/_base.py:117]: Wouldn't [functools.lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache) be faster & easier here? Pretty sure it'll work with a static method. Also pretty sure it's thread-safe.

**Review by dbanty — APPROVED** [review]
Approved with a question/suggestion.

File: rpcq/_base.py (hunk 1 above)
Comment (dbanty) [reply rpcq/_base.py:117]: I think it's thread-safe: https://github.com/python/cpython/blob/3.6/Lib/functools.py#L470

Comment (dbanty) [reply rpcq/_base.py:117]: Also, it was briefly not thread-safe and it was considered a bug which was fixed: https://bugs.python.org/issue28969

Comment (ameyer-rigetti) [reply rpcq/_base.py:115]: OK 👍 Will adjust

**Comment by erichulburd** [comment]
Do you need to patch in VERSION.txt again?

**Comment by ameyer-rigetti** [comment]
@erichulburd I went ahead and bumped the version

**Review by erichulburd — APPROVED** [review]

**Review by notmgsk — APPROVED** [review]
Neat-o!

**Pull request merged**